utils/cmc.go

```diff
@@ -31,8 +43,28 @@ type Response struct {
 	LastUpdate         string `json:"last_updated"`
 }
 
+func init() {
+	pricer = &priceChecker{
+		coins: make(map[string]*coinPrice),
+		mux:   &sync.RWMutex{},
+	}
+}
+
 // RetrieveUsdPrice is used to retrieve the USD price for a coin from CMC
 func RetrieveUsdPrice(coin string) (float64, error) {
+	pricer.mux.RLock()
+	if pricer.coins[coin].price != 0 {
+		if time.Now().After(pricer.coins[coin].nextRefresh) {
+			goto REFRESH
+		}
+		cost := pricer.coins[coin].price
+		pricer.mux.RUnlock()
+		return cost, nil
+	}
+REFRESH:
+	pricer.mux.RUnlock()
+	pricer.mux.Lock()
+	defer pricer.mux.Unlock()
```

Comment:
Add another freshness check here (while mux is write-locked). If two RetrieveUsdPrice of the same coin was called, they could both trigger REFRESH before nextRefresh gets updated.

Reply:
good idea